### PR TITLE
+str #23392 AbstractGraphStageWithMaterializedValue for Java users

### PR DESCRIPTION
**Pull request by ktoso**
(no need to use Tuple2, can be Pair)

Mostly for people who'd panic seeing that we leaked Tuple2 somewhere...
When coding in Java today I felt it was a bit ugly I had to do so, the rest of the API was clean and java friendly

Refs #23392

### CODE REVIEW

File: akka-docs/src/test/java/jdocs/stream/GraphStageDocTest.java [hunk 1]
```diff
@@ -612,8 +614,8 @@ public void onPull() throws Exception {
           });
         }
       };
-
-      return new Tuple2(logic, promise.future());
+      
+      return new Pair<>(logic, FutureConverters.toJava(promise.future()));
```
Comment (ktoso) [inline akka-docs/src/test/java/jdocs/stream/GraphStageDocTest.java]: how did this even compile without the converter?

Comment (johanandren) [inline akka-docs/src/test/java/jdocs/stream/GraphStageDocTest.java]: Not having the diamond in there makes it raw Tuple2 and just gives a warning but compiles?

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 2]
```diff
@@ -24,6 +24,18 @@ import akka.stream.scaladsl.{ GenericGraph, GenericGraphWithChangedAttributes }
 import akka.util.OptionVal
 import akka.annotation.InternalApi
 
+/**
+ * A GraphStage represents a reusable graph stream processing stage.
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Scala API:

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 3]
```diff
@@ -50,9 +62,33 @@ abstract class GraphStageWithMaterializedValue[+S <: Shape, +M] extends Graph[S,
 }
 
 /**
- * A GraphStage represents a reusable graph stream processing stage. A GraphStage consists of a [[Shape]] which describes
- * its input and output ports and a factory function that creates a [[GraphStageLogic]] which implements the processing
- * logic that ties the ports together.
+ * A GraphStage represents a reusable graph stream processing stage.
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Java API:

**Review by johanandren — APPROVED** [review]
LGTM

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 4]
```diff
@@ -50,9 +62,33 @@ abstract class GraphStageWithMaterializedValue[+S <: Shape, +M] extends Graph[S,
 }
 
 /**
- * A GraphStage represents a reusable graph stream processing stage. A GraphStage consists of a [[Shape]] which describes
- * its input and output ports and a factory function that creates a [[GraphStageLogic]] which implements the processing
- * logic that ties the ports together.
+ * Java API: A GraphStage represents a reusable graph stream processing stage.
+ *
+ * Extend this `GraphStageWithMaterializedValue` if you want to provide a materialized value,
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: AbstractGraphStageWithMaterializedValue

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 5]
```diff
@@ -50,9 +62,33 @@ abstract class GraphStageWithMaterializedValue[+S <: Shape, +M] extends Graph[S,
...
+ * represented by the type parameter `M`. If your GraphStage does not need to provide a materialized
+ * value you can instead extende [[GraphStage]] which materializes a [[NotUsed]] value.
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: AbstractGraphStage?

File: akka-docs/src/test/java/jdocs/stream/GraphStageDocTest.java [hunk 6]
```diff
@@ -582,7 +584,7 @@ public void demonstrateAGraphStageWithATimer() throws Exception {
     }
 
     @Override
-    public Tuple2<GraphStageLogic, CompletionStage<A>> createLogicAndMaterializedValue(Attributes inheritedAttributes) {
+    public Pair<GraphStageLogic, CompletionStage<A>> createLogicAndMaterializedValuePair(Attributes inheritedAttributes) {
       Promise<A> promise = Futures.promise();
```
Comment (patriknw) [inline akka-docs/src/test/java/jdocs/stream/GraphStageDocTest.java]: why use scala Promise here? CompletableFuture?

**Review by patriknw — DISMISSED** [review]

File: akka-docs/src/test/java/jdocs/stream/GraphStageDocTest.java (hunk 6 above)
Comment (ktoso) [reply akka-docs/src/test/java/jdocs/stream/GraphStageDocTest.java]: True, will change

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 4 above)
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Thanks

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 5 above)
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Thanks

Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Oh, actually no; we only have one type there :) the handlers have abstract versions

Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: (`abstract` graphstage versions are only needed if we return a pair/tuple; in normal case it's just the logic)

**Pull request merged**